src/lib/datepicker/datepicker.ts

```diff
@@ -0,0 +1,124 @@
+import {
+  Component,
+  TemplateRef,
+  ViewChild,
+  ViewEncapsulation,
+  ChangeDetectionStrategy,
+  ViewContainerRef,
+  Optional,
+  ElementRef, OnDestroy
+} from '@angular/core';
+import {Overlay} from '../core/overlay/overlay';
+import {OverlayRef} from '../core/overlay/overlay-ref';
+import {TemplatePortal} from '../core/portal/portal';
+import {OverlayState} from '../core/overlay/overlay-state';
+import {Dir} from '../core/rtl/dir';
+import {MdError} from '../core/errors/error';
+import {MdDialog} from '../dialog/dialog';
+import {MdDialogRef} from '../dialog/dialog-ref';
+
+
+/** Component responsible for managing the datepicker popup/dialog. */
+@Component({
+  moduleId: module.id,
+  selector: 'md-datepicker',
+  templateUrl: 'datepicker.html',
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class MdDatepicker implements OnDestroy {
+  /** Whether the calendar is in UI is in touch mode. */
+  touchUi: boolean;
+
+  /** The calendar template. */
+  @ViewChild(TemplateRef) calendarTemplate: TemplateRef<any>;
+
+  /** A reference to the overlay when the calendar is opened as a popup. */
+  private _popupRef: OverlayRef;
+
+  /** A reference to the dialog when the calendar is opened as a dialog. */
+  private _dialogRef: MdDialogRef<any>;
+
+  /** A portal containing the calendar for this datepicker. */
+  private _calendarPortal: TemplatePortal;
+
+  /** The input element this datepicker is associated with. */
+  private _inputElementRef: ElementRef;
+
+  constructor(private _dialog: MdDialog, private _overlay: Overlay,
+              private _viewContainerRef: ViewContainerRef, @Optional() private _dir: Dir) {}
+
+  ngOnDestroy() {
+    this.close();
+    this._popupRef.dispose();
```

Comment:
What happens if it's destroyed before you open the popup? Do you need a check here for `this._popupRef`? I think this will break when navigating to another comp in the demo app.